pay/contracts/dropper/HectorDropper.sol

```diff
@@ -0,0 +1,303 @@
+        if (balance > 0) {
+            (bool success, ) = treasury.call{value: balance}('');
+            if (!success) revert FEE_TRANSFER_FAILED();
+        }
+    }
+
+    /* ======== VIEW FUNCTIONS ======== */
+
+    function airdropsByOwner(
+        address from
+    ) external view returns (Airdrop[] memory) {
+        uint256 length = numberOfAirdrops[from];
+        Airdrop[] memory drops = new Airdrop[](length);
+
+        for (uint256 i = 0; i < length; i++) {
+            drops[i] = airdrops[from][i];
+        }
+
+        return drops;
+    }
+
+    function canPerformAirdrop(
+        address from,
+        uint256 index
+    ) external view returns (bool) {
+        if (index >= numberOfAirdrops[from]) return false;
+
+        return airdrops[from][index].releaseTime <= block.timestamp;
+    }
+
+    /* ======== INTERNAL FUNCTIONS ======== */
+
+    function _createAirdrop(
+        address[] calldata tos,
+        uint256 amountPerRecipient,
+        uint48 releaseTime
+    ) internal returns (uint256 index) {
+        uint256 length = tos.length;
+        if (length == 0) revert INVALID_LENGTH();
+        if (amountPerRecipient == 0) revert INVALID_AMOUNT();
+        if (releaseTime <= block.timestamp) revert INVALID_TIME();
+        if (msg.value < fee) revert INSUFFICIENT_FEE();
+
+        for (uint256 i = 0; i < length; i++) {
+            if (tos[i] == address(0)) revert INVALID_ADDRESS();
+        }
+
+        // get the airdrop index
+        index = numberOfAirdrops[msg.sender];
+        numberOfAirdrops[msg.sender] += 1;
+
+        // create a new airdrop
+        Airdrop storage airdrop = airdrops[msg.sender][index];
+        airdrop.from = msg.sender;
+        airdrop.tos = tos;
+        airdrop.amountPerRecipient = amountPerRecipient;
+        airdrop.releaseTime = releaseTime;
+        airdrop.status = AirdropStatus.InProgress;
+        airdrop.fee = fee;
+
+        // fund assets
+        IERC20(token).safeTransferFrom(
+            msg.sender,
+            address(this),
+            amountPerRecipient * length
+        );
+
+        // take fee
+        (bool success, ) = treasury.call{value: fee}('');
+        if (!success) revert FEE_TRANSFER_FAILED();
+    }
+
+    function _cancelAirdrop(uint256 index) internal {
+        if (index >= numberOfAirdrops[msg.sender]) revert INVALID_INDEX();
+
+        Airdrop storage airdrop = airdrops[msg.sender][index];
+        if (airdrop.status != AirdropStatus.InProgress)
+            revert INACTIVE_AIRDROP();
+
+        // update status
+        airdrop.status = AirdropStatus.Cancelled;
+
+        // refund assets
+        IERC20(token).safeTransfer(
+            msg.sender,
+            airdrop.amountPerRecipient * airdrop.tos.length
+        );
+
+        // refund fee
+        (bool success, ) = (msg.sender).call{value: airdrop.fee}('');
+        if (!success) revert FEE_TRANSFER_FAILED();
+    }
+
+    function _releaseAirdrop(address from, uint256 index) internal {
```

Comment:
@bejaxer while i understand the role of "index" variable, wouldn't it be better if we just removed the "index" variable and locate the Airdrop record given the "from" address?

Reply:
in other words, I'd have something like this:
uint256 index = numberOfAirdrops[from];

Reply:
> @bejaxer while i understand the role of "index" variable, wouldn't it be better if we just removed the "index" variable and locate the Airdrop record given the "from" address?

Because one user can create multiple Airdrops